### ISSUE TITLE
Adding masked quote id to QuoteRepository::getList causing infinite loop

### ISSUE THREAD

**Issue by ihor-sviziev** [magento/magento2#33675]
In the Klaviyo extension, we have a plugin that adds a masked quote id to quote search API requests. That looks like quite a small plugin that shouldn't cause any performance or functional damage.

In most cases, it works well, but sometimes we see that such request causing eating 100% CPU by the php-fpm, and Gateway Timeout issues after quite a long time 60 seconds.

Here is a link to the plugin in the Klaviyo extension (it uses API class, which is ok):
https://github.com/klaviyo/magento2-klaviyo/blob/3.0.5/Plugin/Api/CartSearchRepository.php

Detailed investigation showed that issue is coming from the following line:
https://github.com/magento/magento2/blob/afa216974bf1214ee860e7bb9eeaf0474aee1d4f/app/code/Magento/Quote/Model/QuoteIdToMaskedQuoteId.php#L50-L53

-> it goes to https://github.com/magento/magento2/blob/afa216974bf1214ee860e7bb9eeaf0474aee1d4f/app/code/Magento/Quote/Model/QuoteRepository.php#L121
-> and then to 
https://github.com/magento/magento2/blob/7c6b6365a3c099509d6f6e6c306cb1821910aab0/app/code/Magento/Quote/Model/Quote.php#L912-L917

Looks pretty small and simple, right? Yes... but... here we have the recollect totals and re-saving quote
https://github.com/magento/magento2/blob/7c6b6365a3c099509d6f6e6c306cb1821910aab0/app/code/Magento/Quote/Model/Quote.php#L2461-L2468

In case if the quote has `trigger_recollect` = 1:
we running the following code:
https://github.com/magento/magento2/blob/7c6b6365a3c099509d6f6e6c306cb1821910aab0/app/code/Magento/Quote/Model/Quote/TotalsCollector.php#L146-L148

This code triggers an event:
https://github.com/magento/magento2/blob/7c6b6365a3c099509d6f6e6c306cb1821910aab0/app/code/Magento/Quote/Model/Quote/TotalsCollector.php#L263-L264

That execites an event observer:
https://github.com/magento/magento2/blob/7c6b6365a3c099509d6f6e6c306cb1821910aab0/app/code/Magento/SalesRule/etc/events.xml#L30-L32
https://github.com/magento/magento2/blob/7c6b6365a3c099509d6f6e6c306cb1821910aab0/app/code/Magento/SalesRule/Observer/CouponCodeValidation.php#L58-L79
That... runs `$this->cartRepository->getList` method... for which we just added a plugin for adding masked quote ids!

As a result - we have an infinite loop.

### Preconditions (*)
1. Magento 2.3.7 or 2.4-develop

### Steps to reproduce (*)
1. Install Klaviyo extension (tested on version 3.0.5)

    OR Add a plugin, that adds masked quote id to response for the cart search API, for example: 
    - https://github.com/klaviyo/magento2-klaviyo/blob/3.0.5/etc/webapi_rest/di.xml#L3-L5
    -  https://github.com/klaviyo/magento2-klaviyo/blob/3.0.5/Plugin/Api/CartSearchRepository.php
    
2. Create Cart Price Rule with coupon (let's give some 10% discount)
3. Add some product to the shopping cart, apply the coupon from step 2
4. Run the following code with and w/o Klaviyo extension:  
```bash
curl --location --request GET 'https://www.mymagento.com/index.php/rest/default/V1/carts/search?searchCriteria%5Bfilter_groups%5D%5B0%5D%5Bfilters%5D%5B0%5D%5Bfield%5D=updated_at&searchCriteria%5Bfilter_groups%5D%5B0%5D%5Bfilters%5D%5B0%5D%5Bvalue%5D=2021-06-11+11%3A38%3A16&searchCriteria%5Bfilter_groups%5D%5B0%5D%5Bfilters%5D%5B0%5D%5Bcondition_type%5D=gt&searchCriteria%5BsortOrders%5D%5B0%5D%5Bfield%5D=updated_at&searchCriteria%5BsortOrders%5D%5B0%5D%5Bdirection%5D=DESC&searchCriteria%5BcurrentPage%5D=1&searchCriteria%5BpageSize%5D=50' \
--header 'Authorization: Bearer dk6zp483qmck8omd7sov1r3b9is7kxh'
```

  Note: here is documentation on how to get access token https://devdocs.magento.com/guides/v2.4/get-started/authentication/gs-authentication-token.html#request-token
    ✔ we successfully retrieved the list of carts

5. Go to admin, edit the Cart Price Rule that you created in step 2, do the following steps:
    - change the coupon code, save the rule
    - disable the cart price rule, save it
6. Run the code from step 4

### Expected result (*)
1. You should get the JSON with a list of carts

### Actual result (*)
1. ❌We have 504 Gateway Time-out error 
![image](https://user-images.githubusercontent.com/1873745/128130473-1fe775be-652e-487d-805e-a33975d42ddb.png)

2. 

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.4/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [x] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

**Comment by jsamhall** [comment]
Chiming in as this issue has impacted our production environment, and becomes somewhat severe when the payment processor relies on REST API to update orders. In our case, we use Bolt for payment processing and it uses webhooks / Magento REST API to manage order status according to payment status at the processor.

In our case, we have become inundated with "Request Timeout" from Bolt, but it is of course due to the issue described here.

I have resolved the issue by patching the Klaviyo CartSearchRepository plugin and exiting before it queries the QuoteRepository a second time. I have achieved this by caching imploded array keys of the passed CartSearchResults into a static property on the Plugin and checking that before querying the QuoteRepository.

My relevant changes here:
```
/**
     * Identification of CartSearchResults being processed; part of fix to avoid infinite loops
     * @var array 
     */
    private static $processingSearchResults = [];

    public function __construct(
        CartExtensionFactory $extensionFactory,
        QuoteIdToMaskedQuoteId $quoteIdToMaskedQuoteId
    ){...}
    /**
     * Add "kl_masked_id" extension attribute to order data object to make it accessible in API data
     *
     * @param CartRepositoryInterface $subject
     * @param CartSearchResults $searchResult
     *
     * @return CartSearchResults
     */
    public function afterGetList(CartRepositoryInterface $subject, CartSearchResults $searchResult)
    {
        $quotes = $searchResult->getItems();

        // Prevent infinite loop as Magento CouponValidator may invoke CartRepository::getList if Quote has a Coupon applied
        // @see https://github.com/magento/magento2/issues/33675
        if($this->isProcessingCartSearchResults($searchResult)){
            return $searchResult;
        }

        foreach ($quotes as $quote) {
            $maskedId = $this->getMaskedIdFromQuoteId($quote->getId());
            $extensionAttributes = $quote->getExtensionAttributes();
            $extensionAttributes = $extensionAttributes ? $extensionAttributes : $this->extensionFactory->create();
            $extensionAttributes->setData(self::KL_MASKED_ID, $maskedId);
            $quote->setExtensionAttributes($extensionAttributes);
        }

        return $searchResult;
    }

    /**
     * Evaluate if this CartSearchResults object has already been handled
     * @param CartSearchResults $searchResults
     * @return bool
     */
    private function isProcessingCartSearchResults(CartSearchResults $searchResults){
        $processingKey = implode('_', array_keys($searchResults->getItems()));
        if(in_array($processingKey, self::$processingSearchResults)){
            return true;
        }

        self::$processingSearchResults[] = $processingKey;
        return false;
    }
```
 
@ihor-sviziev It appears you are the author of the Klaviyo Magento 2 integration? May you provide any input to my patch?

**Comment by ihor-sviziev** [comment]
@jsamhall, I'm not the author of the klaviyo extension, but looks like this issue is fixed in version 3.0.8

https://github.com/klaviyo/magento2-klaviyo/compare/3.0.7...3.0.8

**Comment by jsamhall** [comment]
@ihor-sviziev  thank you very much for the thoughtful response! I sincerely appreciate it! Cheers

**Comment by sidolov** [comment]
Hi @ihor-sviziev. Thank you for your report. 
The issue has been fixed in magento/magento2#33324 by @ihor-sviziev in 2.4-develop branch
Related commit(s):
 - [b53d2afcf6c23e3679729bc0f5c3ac4d81d2844b](https://github.com/magento/magento2/commit/b53d2afcf6c23e3679729bc0f5c3ac4d81d2844b)

The fix will be available with the upcoming release.
